src/core/file_discard_queue.cc

```diff
@@ -0,0 +1,300 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2024 ScyllaDB
+ */
+
+#ifdef SEASTAR_MODULE
+module;
+#endif
+
+#include <algorithm>
+#include <cassert>
+#include <chrono>
+#include <stdexcept>
+
+#include <fmt/format.h>
+
+#ifdef SEASTAR_MODULE
+module seastar;
+#else
+#include <seastar/core/do_with.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/file_discard_queue.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/when_all.hh>
+#endif
+
+namespace seastar {
+
+using file_discard_context = internal::file_discard_context;
+using file_discard_phase = internal::file_discard_phase;
+
+file_discard_queue::file_discard_queue(const file_discard_queue_config& config)
+    : _config{config}
+    , _available_requests{config._requests_per_second}
+    , _refill_available_requests_timer{[this] { refill_available_requests(); }}
+{
+    if ((config._discard_block_size == 0u) || (config._discard_block_size % 4096u != 0u)) {
+        throw std::invalid_argument{"Cannot construct file_discard_queue: block size must be positive and aligned to 4KiB!"};
+    }
+
+    if (_config._requests_per_second == 0u) {
+        throw std::invalid_argument{"Cannot construct file_discard_queue: RPS must be positive!"};
+    }
+
+    _currently_discarded_files.reserve(std::max(4u * _config._requests_per_second, uint64_t{256u}));
+    _refill_available_requests_timer.arm_periodic(std::chrono::seconds{1});
```

Comment:
Consider you have 1M requests per second. AFAIU it means, that the discarder will emit 1M requests almost (there's magic 8u constant in try_discard_blocks) instantly and then will be idling for the remaining (almost) second.

Reply:
That's true. Also, if the timer expires right after the requests were issued, then another 1M will be issued. When using periodic timer such situation may happen. In such case the bandwidth at that second may be almost twice as big as the configured one. It is wrong.

During creation of `file_discard_queue_global_test.cc` I also had impression that setting the periodic timer in constructor is problematic. However, I wanted to push the draft ASAP to get some general feedback about used approach. Also, I see some corner cases in other parts of the PR that need to be discussed.

Regarding the timer -  currently I have two ideas:

- decrease the period of the timer and refill only fraction of the available shares e.g. given `1/n` s period refill `RPS/n` shares - it could alleviate "long idling" problem, but still if the refill happens right after the first portion of discards was issued, then the bandwidth for the first second may be larger than the configured one (by BW/n)
- do not use periodic timer - instead arm timer after issuing discards that caused decreasing number of shares

Regarding the magic constant - `file_discard_queue` processes files in FIFO order. If any file was opened, then the queue tries to discard that file completely before opening next ones. My intention was to not open too many files at once. `8` is just an arbitrary value.

Reply:
You're trying to implement token-bucket algorithm, it can be done without a timer and with smoother refill. All you need is to refill tokens not by timer, but synchronously, every time you try to grab tokens from them. For that, there should a function `tokens = fn(duration<double>)` that converts duration into tokens. Every time you grab tokens you first put some of them back using that fn() and note the time of the replenish to calculate duration for the next refill. There's util/shared_token_bucket.hh that implements it, but it's node-wide, not shard-local and thus has some extra complexity due to that.